{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'adjustText'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-49d4c423f2a4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mnumpy\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mscipy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstats\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0;32mfrom\u001b[0m \u001b[0madjustText\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0madjust_text\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'adjustText'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import scipy.stats\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Initializing DepMap database classes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DepMap:\n",
    "    \"\"\"\n",
    "    This class allows for interaction with the data found in the Achilles and\n",
    "    DEMETER2 data set from the public DepMap portal. \n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        self.load_achilles()\n",
    "        self.load_rnai()\n",
    "        self.load_guides()\n",
    "        self.load_mutations()\n",
    "\n",
    "    # DepMap dataset load in\n",
    "    def load_achilles(self, achilles_path=\"../data/depmap/20Q1/202020226_Achilles_gene_effect.csv\"):\n",
    "        \"\"\"\n",
    "        This function loads in the Achilles data. The path to data file is\n",
    "        optional and only required if there is an updated or older version of\n",
    "        the data to be used. That is the only time this function needs to be\n",
    "        manually called, otherwise it will be called during object\n",
    "        initialization.\n",
    "        \n",
    "        :param achilles_path: (str) path to Achilles data set.\n",
    "\n",
    "        :return: (None) Initializes raw_achilles and achilles attributes.\n",
    "        \"\"\"\n",
    "        self.achilles = Achilles(achilles_path)\n",
    "\n",
    "    def load_rnai(self, rnai_path=\"../data/depmap/20Q1/20200226_D2_combined_gene_dep_scores.csv\"):\n",
    "        \"\"\"\n",
    "        This function loads in the RNAi (DEMETER2) data. The path to data file is\n",
    "        optional and only required if there is an updated or older version of\n",
    "        the data to be used. That is the only time this function needs to be\n",
    "        manually called, otherwise it will be called during object\n",
    "        initialization.\n",
    "        \n",
    "        :param rnai_path: (str) path to RNAi data set.\n",
    "\n",
    "        :return: (None) Initializes raw_rnai and rnai attributes.\n",
    "        \"\"\"\n",
    "        self.rnai = RNAi(rnai_path)\n",
    "\n",
    "    def load_guides(self, guide_path=\"../data/depmap/20Q1/20200304_Achilles_guide_map.csv\"):\n",
    "        \"\"\"\n",
    "        This function loads in the CRISPR guide data. The path to data file is\n",
    "        optional and only required if there is an updated or older version of\n",
    "        the data to be used. That is the only time this function needs to be\n",
    "        manually called, otherwise it will be called during object\n",
    "        initialization.\n",
    "        \n",
    "        :param guide_path: (str) path to CRISPR guide data set.\n",
    "\n",
    "        :return: (None) Initializes guide attributes.\n",
    "        \"\"\"\n",
    "        self.guides = CRISPRGuides(guide_path)\n",
    "\n",
    "    def load_mutations(self, mutation_path=\"../data/depmap/20Q1/20200309_CCLE_mutations.csv\"):\n",
    "        \"\"\"\n",
    "        This function loads in the DepMap mutation data. The path to data file is\n",
    "        optional and only required if there is an updated or older version of\n",
    "        the data to be used. That is the only time this function needs to be\n",
    "        manually called, otherwise it will be called during object\n",
    "        initialization.\n",
    "\n",
    "        :param mutation_path: (str) path to DepMap mutation data.\n",
    "\n",
    "        :return: (None) Initializes mutation attribute.\n",
    "        \"\"\"\n",
    "        self.mutations = Mutation(mutation_path)\n",
    "\n",
    "    # Plotting functions\n",
    "    def plot_distributions(self, gene_list, save=False):\n",
    "        for gene in gene_list:\n",
    "            show = not save\n",
    "            self._make_distribution(gene, show=show)\n",
    "\n",
    "    def plot_alignments(self, thr=5, show=True):\n",
    "        self._make_alignments_ceres(thr, show)\n",
    "\n",
    "    def _make_alignments_ceres(self, thr, show=True):\n",
    "        \"\"\"\n",
    "        This function plots the mean CERES score from all cell lines for each\n",
    "        gene on the x-axis and the mean number of alignments of the guide\n",
    "        targeting that specific gene. It colors and labels all points with the\n",
    "        number of alignments greater than or equal to the provided threshold\n",
    "        that also have a CERES score less than or equal to -0.5.\n",
    "\n",
    "        :params thr: (int) The threshold for the number of alignments above\n",
    "            which (inclusive) to color and label.\n",
    "        :params show: (bool) default is True. Whether to show the plot or to\n",
    "            save it. It saves the plot to ../output/{hlp.insert_date()}/figures/DepMap/\n",
    "\n",
    "        :return: (None). Shows plot or saves to indicated path.\n",
    "        \"\"\"\n",
    "        # Averaging number of alignments per gene\n",
    "        alignments = self.guides.guides_data[\"n_alignments\"].astype(\"int\")\n",
    "        alignments = alignments.groupby(\"gene\").aggregate(np.mean).sort_values()\n",
    "        \n",
    "        # Avering CERES score per gene\n",
    "        achilles_avg = self.achilles.achilles_data.mean(axis=0)\n",
    "        \n",
    "        # Only selecting genes which are in both data sets\n",
    "        common_genes = set(achilles_avg.index).intersection(set(alignments.index))\n",
    "        achilles_avg = achilles_avg[common_genes].sort_index()\n",
    "        alignments = alignments[common_genes].sort_index()\n",
    "\n",
    "        # Filtering based on threshold and CERES score\n",
    "        selected_genes = set(alignments[alignments >= thr].index).intersection(\n",
    "            set(achilles_avg[achilles_avg <= -0.5].index))\n",
    "        selected_indices = alignments.index.isin(selected_genes)\n",
    "        \n",
    "        # Saving genes below threshold for cuts and below -0.5 for dependency\n",
    "        target_genes = set(alignments[alignments < thr].index).intersection(\n",
    "            set(achilles_avg[achilles_avg <= -0.5].index))\n",
    "        with open(\"../code/tmp/06_Dependent_LowCuts.csv\", \"w\") as f:\n",
    "            f.write(','.join(list(target_genes)))\n",
    "            \n",
    "        # Plotting\n",
    "        fig, ax = plt.subplots()\n",
    "        ax.scatter(achilles_avg.values[selected_indices], \n",
    "                   alignments.values[selected_indices], s=1, color=\"r\")\n",
    "        ax.scatter(achilles_avg.values[~selected_indices],\n",
    "                   alignments.values[~selected_indices], s=1)\n",
    "        \n",
    "        for index in [i for i, x in enumerate(selected_indices) if x]:\n",
    "            ax.annotate(alignments.index[index], \n",
    "                        (achilles_avg.values[index], alignments.values[index]), \n",
    "                        fontsize=4)\n",
    "        plt.xlim(min(achilles_avg.values)-0.05, 0)\n",
    "        plt.xlabel(\"Gene Effect\")\n",
    "        plt.ylabel(\"Mean Number of Guide Alignments\")\n",
    "\n",
    "        if show:\n",
    "            plt.show()\n",
    "        else:\n",
    "            if not os.path.exists(f\"../output/{hlp.insert_date()}/figures/DepMap/\"):\n",
    "                os.path.makedirs(f\"../output/{hlp.insert_date()}/figures/DepMap/\")\n",
    "            plt.savefig(f\"../output/{hlp.insert_date()}/figures/DepMap/{hlp.insert_date()}-GuideAlignments-CERES-thr-{thr}.pdf\")\n",
    "            plt.close()\n",
    "\n",
    "    def _make_distribution(self, gene, show=True):\n",
    "        \"\"\"\n",
    "        This function plots distributions of the Achilles and RNAi data similar\n",
    "        to those provided on the DepMap portal. \n",
    "\n",
    "        :params gene: (str) the gene name to make the distribution plot for.\n",
    "        :params show: (bool) default is True. Whether to show the plot or to\n",
    "            save it. It saves the plot to ../output/{hlp.insert_date()}/figures/DepMap/\n",
    "        \n",
    "        :return: (None). Shows plot or saves to indicaetd path.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            sns.distplot(self.achilles.achilles_data[gene], hist=False, color=\"#2fa3de\", \n",
    "                        kde_kws={\"shade\": True, \"color\": \"#2fa3de\"}, label=\"Achilles\")\n",
    "        except KeyError:\n",
    "            print(f\"\\t\\tData for gene symbol {gene} not in Achilles data.\")\n",
    "\n",
    "        try:\n",
    "            sns.distplot(self.rnai.rnai_data[gene], hist=False, color=\"#a86ce0\", \n",
    "                        kde_kws={\"shade\": True, \"color\": \"#a86ce0\"}, label=\"RNAi\")\n",
    "        except KeyError:\n",
    "            print(f\"\\t\\tData for gene symbol {gene} not in RNAi data.\")\n",
    "\n",
    "        yrange = plt.gca().get_ylim()\n",
    "        plt.vlines([-1, 0], ymin=0, ymax=yrange[1], colors=[\"r\", \"k\"],\n",
    "                linestyles=[\"dashed\", \"solid\"])\n",
    "        plt.yticks([])\n",
    "        plt.xlabel(\"Gene Effect\")\n",
    "        if show:\n",
    "            plt.title(gene)\n",
    "            plt.show()\n",
    "        else:\n",
    "            if not os.path.exists(f\"../output/{hlp.insert_date()}/figures/DepMap/Dependency-Distributions/\"):\n",
    "                os.makedirs(\n",
    "                    f\"../output/{hlp.insert_date()}/figures/DepMap/Dependency-Distributions/\")\n",
    "            plt.savefig(\n",
    "                f\"../output/{hlp.insert_date()}/figures/DepMap/Dependency-Distributions/{hlp.insert_date()}-DepMapPlot-{gene}.pdf\")\n",
    "            plt.close()\n",
    "\n",
    "        return None\n",
    "\n",
    "\n",
    "class Mutation:\n",
    "    \"\"\"\n",
    "    This class allows for easy interaction with and interrogation of the CCLE\n",
    "    mutation data set.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, mutation_path):\n",
    "        self._load_mutations(mutation_path)\n",
    "        self._top_mutations = None\n",
    "\n",
    "    def _load_mutations(self, mutation_path):\n",
    "        \"\"\"\n",
    "        This function loads in the DepMap mutation data. The path to data file is\n",
    "        optional and only required if there is an updated or older version of\n",
    "        the data to be used. That is the only time this function needs to be\n",
    "        manually called, otherwise it will be called during object\n",
    "        initialization.\n",
    "\n",
    "        :param mutation_path: (str) path to DepMap mutation data.\n",
    "\n",
    "        :return: (None) Initializes mutation attributes.\n",
    "        \"\"\"\n",
    "        self.mutations_data = pd.read_csv(mutation_path)\n",
    "        self.mutations_data.set_index(\"Unnamed: 0\", inplace=True)\n",
    "        # Safer data extraction\n",
    "        self.mutations_data.fillna(value=False, inplace=True)\n",
    "        return None\n",
    "\n",
    "    def cell_line_extract(self, gene, mutation_class=False, mutation_type=False, tcga_count=0,\n",
    "                          cosmic_count=0, deleterious=False, tcga_hotspot=False, cosmic_hotspot=False):\n",
    "        \"\"\"\n",
    "        This function returns the cell line IDs which have a mutation in the\n",
    "        given gene. This result can be further filtered by the specific optional\n",
    "        parameters.\n",
    "\n",
    "        :param gene: (str) Hugo_Symbol of gene name to be searched.\n",
    "        :param mutation_class: (str) optional. Return cell lines with a mutation\n",
    "            in the provided gene with this classification. \n",
    "            Classes are: \n",
    "            ['Silent', 'Missense_Mutation', 'In_Frame_Del', 'Nonsense_Mutation',\n",
    "            'Frame_Shift_Ins', 'Splice_Site', 'Intron','De_novo_Start_OutOfFrame', 'Frame_Shift_Del', 'Start_Codon_SNP', 'In_Frame_Ins', 'Nonstop_Mutation', \"5'UTR\", \"5'Flank\", 'Stop_Codon_Del', 'Start_Codon_Del', 'IGR', 'Stop_Codon_Ins',\n",
    "            'Start_Codon_Ins', \"3'UTR\"]\n",
    "        :param mutation_type: (str) optional. Return cell lines with a mutation\n",
    "            in the provided gene with this mutation type.\n",
    "            Options are: ['SNP', 'DEL', 'INS', 'DNP', 'TNP']\n",
    "        :param tcga_count: (int) optional. Return cell lines with a mutation in\n",
    "            the provided gene with at least this number of times this mutation is observed in\n",
    "            TCGA.\n",
    "        :param cosmic_count: (int) optional. Return cell lines with a\n",
    "            the provided gene with at least this number of times this mutation\n",
    "            is observed in COSMIC.\n",
    "        :param deleterious: (bool) optional. Return cell lines with a\n",
    "            mutation in the provided gene only if the the mutation is deleterious.\n",
    "        :param tcga_hotspot: (bool) optional. Return cell lines with a\n",
    "            mutation in the provided gene only if this mutation is commonly\n",
    "            found in TCGA.\n",
    "        :param cosmic_hotspot: (bool) optional. Return cell lines with a\n",
    "            mutation in the provided gene only if this mutation is commonly\n",
    "            found in COSMIC.\n",
    "\n",
    "        :return: (List) A list of cell lines with a mutation in the provided\n",
    "            gene meeting all of the different parameter options provided.\n",
    "        \"\"\"\n",
    "        filtered_mutations = self.mutations_data[self.mutations_data[\"Hugo_Symbol\"] == gene]\n",
    "        \n",
    "        # Matching function parameters to DepMap mutation table columns and\n",
    "        # parameter value. Using ordered structure of dict.\n",
    "        parameters = {\"Variant_Classification\": mutation_class,\n",
    "                      \"Variant_Type\": mutation_type, \n",
    "                      \"TCGAhsCnt\": tcga_count,\n",
    "                      \"COSMIChsCnt\": cosmic_count,\n",
    "                      \"isDeleterious\": deleterious,\n",
    "                      \"isTCGAhotspot\": tcga_hotspot,\n",
    "                      \"isCOSMIChotspot\": cosmic_hotspot}\n",
    "\n",
    "        # Need to do below to allow for filtering only when user supplies a\n",
    "        # parameter.\n",
    "        filter_conditions = np.array(list(parameters.keys()))\n",
    "        fil = np.array([mutation_class, mutation_type, bool(tcga_count), bool(cosmic_count),\n",
    "                        deleterious, tcga_hotspot, cosmic_hotspot])\n",
    "        filter_conditions = filter_conditions[fil]\n",
    "\n",
    "        # Iterating over user selected parameters and filtering data\n",
    "        for column in filter_conditions:\n",
    "            if column.startswith(\"is\"):  # Boolean Filters\n",
    "                filtered_mutations = filtered_mutations[filtered_mutations[column]]\n",
    "            elif column.endswith(\"Cnt\"):  # Count Filters\n",
    "                filtered_mutations = \\\n",
    "                    filtered_mutations[filtered_mutations[column]\n",
    "                                       >= parameters[column]]\n",
    "            else:  # String Matching\n",
    "                filtered_mutations = filtered_mutations[filtered_mutations[column]\n",
    "                                                        == parameters[column]]\n",
    "        return filtered_mutations\n",
    "\n",
    "    def top_mutations(self):\n",
    "        # Prevents redoing calculation if already requested\n",
    "        if not self._top_mutations:\n",
    "            self._get_top_mutations()\n",
    "            \n",
    "        return self._top_mutations\n",
    "    \n",
    "    def _get_top_mutations(self):\n",
    "        \"\"\"\n",
    "        This function returns the top mutations in the mutations data set. It\n",
    "        pre-filters the list of all genes to be TCGA hotspots to counter act the\n",
    "        effect of longer genes having more mutations naturally. It is assumed\n",
    "        that TCGA already accounts for this difference.\n",
    "        \"\"\"\n",
    "        self._top_mutations = self.mutations_data[self.mutations_data[\"isTCGAhotspot\"]]\n",
    "        self._top_mutations = self._top_mutations.groupby(\"Hugo_Symbol\").count()\n",
    "        self._top_mutations = self._top_mutations.sort_values(\"DepMap_ID\",\n",
    "                                                              ascending=False)\n",
    "        \n",
    "        # Want it to be immutable since it's already sorted\n",
    "        self._top_mutations = tuple(zip(tuple(self._top_mutations.index), \n",
    "                                  tuple(self._top_mutations[\"DepMap_ID\"])))\n",
    "        return None\n",
    "\n",
    "\n",
    "class RNAi:\n",
    "    \"\"\"\n",
    "    This class allows for interaction with and interrogation of the RNAi\n",
    "    DEMETER2 dataset.\n",
    "    \"\"\"\n",
    "    def __init__(self, rnai_path):\n",
    "        self._load_rnai(rnai_path)\n",
    "\n",
    "    def _load_rnai(self, rnai_path):\n",
    "        \"\"\"\n",
    "        This function loads in the RNAi (DEMETER2) data. The path to data file is\n",
    "        optional and only required if there is an updated or older version of\n",
    "        the data to be used. That is the only time this function needs to be\n",
    "        manually called, otherwise it will be called during object\n",
    "        initialization.\n",
    "\n",
    "        :param rnai_path: (str) path to RNAi data set.\n",
    "\n",
    "        :return: (None) Initializes raw_rnai and rnai attributes.\n",
    "        \"\"\"\n",
    "        self.raw_rnai = pd.read_csv(rnai_path)\n",
    "        self._reformat_rnai_data()\n",
    "        return None\n",
    "\n",
    "    def _reformat_rnai_data(self):\n",
    "        \"\"\"\n",
    "        This function reformats the raw RNAi data. It reindexes the\n",
    "        data by the Tissue ID and then fixes the gene names to remove extraneous\n",
    "        symbols.\n",
    "        \n",
    "        :return: (None) Sets self.rnai to a reindexed and cleaned up version of the raw\n",
    "            data for use in downstream analysis.\n",
    "        \"\"\"\n",
    "        updated_data = self.raw_rnai.set_index(\"Unnamed: 0\")\n",
    "        updated_data = updated_data.transpose()\n",
    "        updated_data.columns = updated_data.columns.str.replace(\n",
    "            \"\\s\\(\\\\d+\\)\", \"\", regex=True)\n",
    "        self.rnai_data = updated_data\n",
    "        return None\n",
    "\n",
    "\n",
    "class Achilles:\n",
    "    \"\"\"\n",
    "    This class allows for interaction with and interrogation of the CRISPR\n",
    "    Achilles dataset.\n",
    "    \"\"\"\n",
    "    def __init__(self, achilles_path):\n",
    "        self._load_achilles(achilles_path)\n",
    "\n",
    "    def _load_achilles(self, achilles_path):\n",
    "        \"\"\"\n",
    "        This function loads in the Achilles data. The path to data file is\n",
    "        optional and only required if there is an updated or older version of\n",
    "        the data to be used. That is the only time this function needs to be\n",
    "        manually called, otherwise it will be called during object\n",
    "        initialization.\n",
    "\n",
    "        :param achilles_path: (str) path to Achilles data set.\n",
    "\n",
    "        :return: (None) Initializes raw_achilles and achilles attributes.\n",
    "        \"\"\"\n",
    "        self.raw_achilles = pd.read_csv(achilles_path)\n",
    "        self._reformat_achilles_data()\n",
    "        return None\n",
    "\n",
    "\n",
    "    def _reformat_achilles_data(self):\n",
    "        \"\"\"\n",
    "        This function reformats the raw dependency data so that information\n",
    "        regarding tissue specific dependencies can be extracted. It reindexes the\n",
    "        data by the DepMap_ID and then fixes the gene names to remove extraneous\n",
    "        symbols.\n",
    "\n",
    "        :return: (None) sets self.achilles to a reindexed and cleaned up version of the raw\n",
    "            dependency probability data for use in downstream analysis.\n",
    "        \"\"\"\n",
    "        dependency_data_update = self.raw_achilles.rename(\n",
    "            columns={\"Unnamed: 0\": \"DepMap_ID\"})\n",
    "        dependency_data_update = dependency_data_update.set_index(\n",
    "            \"DepMap_ID\")\n",
    "\n",
    "        dependency_data_update.columns = dependency_data_update.columns.str.replace(\n",
    "            \" \\(\\\\d+\\)\", \"\", regex=True)\n",
    "\n",
    "        self.achilles_data = dependency_data_update\n",
    "        return None\n",
    "\n",
    "\n",
    "class CRISPRGuides:\n",
    "    \"\"\"\n",
    "    This class allows for interaction with and interrogation of the CRISPR\n",
    "    Guides dataset.\n",
    "    \"\"\"\n",
    "    def __init__(self, guides_path):\n",
    "        self._load_guides(guides_path)\n",
    "\n",
    "    def _load_guides(self, guide_path=\"../data/depmap/20Q1/20200304_Achilles_guide_map.csv\"):\n",
    "        \"\"\"\n",
    "        This function loads in the CRISPR guide data. The path to data file is\n",
    "        optional and only required if there is an updated or older version of\n",
    "        the data to be used. That is the only time this function needs to be\n",
    "        manually called, otherwise it will be called during object\n",
    "        initialization.\n",
    "        \n",
    "        :param guide_path: (str) path to CRISPR guide data set.\n",
    "\n",
    "        :return: (None) Initializes guide attributes.\n",
    "        \"\"\"\n",
    "        self.raw_guides = pd.read_csv(guide_path)\n",
    "        self._reformat_guides()\n",
    "\n",
    "    def _reformat_guides(self):\n",
    "        \"\"\"\n",
    "        This function reformats the raw guides data. It reindexes the\n",
    "        data by the gene and also fixes the gene names to remove extraneous\n",
    "        symbols.\n",
    "        \n",
    "        :return: (None) Sets self.guides to a reindexed and cleaned up version of the raw\n",
    "            data for use in downstream analysis.\n",
    "        \"\"\"\n",
    "        updated_guides = self.raw_guides.set_index(\"gene\")\n",
    "        updated_guides = updated_guides.transpose()\n",
    "        updated_guides.columns = updated_guides.columns.str.replace(\n",
    "            \" \\(\\\\d+\\)\", \"\", regex=True)\n",
    "        updated_guides = updated_guides.transpose()\n",
    "        self.guides_data = updated_guides.sort_index()\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:11: DtypeWarning: Columns (20,30,31) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  # This is added back by InteractiveShellApp.init_path()\n"
     ]
    }
   ],
   "source": [
    "depmap = DepMap()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_info = pd.read_csv(\"../data/depmap/20Q1/20200227_sample_info_public.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mean CERES Score Difference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras_mut = depmap.mutations.mutations_data[depmap.mutations.mutations_data[\"Hugo_Symbol\"] == \"KRAS\"][\"DepMap_ID\"].unique().tolist()\n",
    "kras_wt = depmap.mutations.mutations_data[depmap.mutations.mutations_data[\"Hugo_Symbol\"] != \"KRAS\"][\"DepMap_ID\"].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:1418: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n",
      "  return self._getitem_tuple(key)\n"
     ]
    }
   ],
   "source": [
    "kras_mut_achilles = depmap.achilles.achilles_data.loc[kras_mut, :].dropna()\n",
    "kras_wt_achilles = depmap.achilles.achilles_data.loc[kras_wt, :].dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_ceres_mut = kras_mut_achilles.mean(axis=0)\n",
    "mean_ceres_wt = kras_wt_achilles.mean(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_results = scipy.stats.ttest_ind(kras_mut_achilles, kras_wt_achilles, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_difference = mean_ceres_mut - mean_ceres_wt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = {\"genes\": mean_difference.index, \"differences\": mean_difference.values, \"logp\": -np.log10(test_results[1])}\n",
    "\n",
    "plot_df = pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df.to_csv(\"../output/KRAS_Mut_Wt_Group_Differences_Volcano.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Making Plot with Broken Axes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First need to extract only negative differences for plot and make these values the absolute value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>differences</th>\n",
       "      <th>logp</th>\n",
       "      <th>color</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>0.002613</td>\n",
       "      <td>0.098647</td>\n",
       "      <td>k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>A2ML1</td>\n",
       "      <td>0.014006</td>\n",
       "      <td>0.679085</td>\n",
       "      <td>k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>A4GNT</td>\n",
       "      <td>0.004531</td>\n",
       "      <td>0.148413</td>\n",
       "      <td>k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>0.007392</td>\n",
       "      <td>0.201896</td>\n",
       "      <td>k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>AADAC</td>\n",
       "      <td>0.022262</td>\n",
       "      <td>0.944062</td>\n",
       "      <td>k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18314</td>\n",
       "      <td>ZSWIM2</td>\n",
       "      <td>0.011716</td>\n",
       "      <td>0.471704</td>\n",
       "      <td>k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18319</td>\n",
       "      <td>ZSWIM7</td>\n",
       "      <td>0.041437</td>\n",
       "      <td>2.244959</td>\n",
       "      <td>r</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18321</td>\n",
       "      <td>ZUP1</td>\n",
       "      <td>0.008913</td>\n",
       "      <td>0.373416</td>\n",
       "      <td>k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18328</td>\n",
       "      <td>ZYG11A</td>\n",
       "      <td>0.006189</td>\n",
       "      <td>0.228736</td>\n",
       "      <td>k</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>18329</td>\n",
       "      <td>ZYG11B</td>\n",
       "      <td>0.014779</td>\n",
       "      <td>0.611489</td>\n",
       "      <td>k</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9320 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        genes  differences      logp color\n",
       "2         A2M     0.002613  0.098647     k\n",
       "3       A2ML1     0.014006  0.679085     k\n",
       "6       A4GNT     0.004531  0.148413     k\n",
       "7        AAAS     0.007392  0.201896     k\n",
       "9       AADAC     0.022262  0.944062     k\n",
       "...       ...          ...       ...   ...\n",
       "18314  ZSWIM2     0.011716  0.471704     k\n",
       "18319  ZSWIM7     0.041437  2.244959     r\n",
       "18321    ZUP1     0.008913  0.373416     k\n",
       "18328  ZYG11A     0.006189  0.228736     k\n",
       "18329  ZYG11B     0.014779  0.611489     k\n",
       "\n",
       "[9320 rows x 4 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plot_negative = plot_df[plot_df[\"differences\"] < 0]\n",
    "plot_negative[\"differences\"] = abs(plot_negative[\"differences\"])\n",
    "plot_negative[\"color\"] = ['r' if x >= 1.301 else 'k' for x in plot_negative['logp']]\n",
    "plot_negative"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:64: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:69: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:74: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:79: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:84: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:89: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:94: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:99: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:104: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:109: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:114: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:119: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n",
      "/Users/kevinkapner/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:124: UserWarning: You have used the `textcoords` kwarg, but not the `xytext` kwarg.  This can lead to surprising results.\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYYAAAECCAYAAAD6oXArAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nOzdd3RUVdfA4d9JL4QSaiihKuWVagCBCAkEEAQURCxIkxhBUVFAUJD2gljwRVAEKYpfAFFBOtKiEQFFuo0iSBcIgYSWkDJzvj9OMmSSkARSJsB+1mJl5s4t547Lu+e0fZTWGiGEECKVk6MLIIQQonCRwCCEEMKOBAYhhBB2JDAIIYSwI4FBCCGEHQkMQggh7Nx1gUEpFeboMuTW7X4PUn7Hut3LD7f/PRT28quCnseglKoNvAKUAiKA34EngWTgXa316Xy+/g6tdUB+XiO/3e73IOV3rNu9/HD730NhL79LQV9Qa70PGKCUcgJmA22Aw4AFuFDQ5RFCiNxQSvkBbQH3mzislFLquRzsFwWs0lpbbqlwt6jAawwASqkuwAjgY2AcUBtoD5TVWn+Wyf5hQBiAt7f3/bVq1brla587d47SpUvf8vGFwe1+D1J+x7rdyw+F5x4SExM5cOAAXl5euLjk/Hd2fHw8np6eWe6jtSYuLg4vLy+qVKmSy5LCzp07o7XWOfrSHBIYbBdXajVg0Vp3UUo1BxprradmdUxAQIDesWNHwRRQCCGy8Pnnn7N27Vq++uqrfDl/XFwcxYsX58qVK7i5ueXqXEqpnTltvirwzmelVJBSappS6lNgDTBfKTUDGAgsLOjyCCHErUpISKBEiRIAREZGMmrUKACWLFnCY489RunSpQkODqZly5YcOnTIdtxPP/1E8eLFSUxMBCA5OZmnn36aVq1a8cADD7By5UoAvLy8cHV1JSkpqUDvyxF9DJFAZLrNXxd0OYQQIj9s2bKF6dOnM336dCZOnMj8+fPZvHkzM2fOZPLkyQB8++23dO/enYiICDp06MDatWupXr06CxcuRGvNxYsXHXoPd91wVSGEyC8HDhxg6NChfPPNN3Z9CLGxsXb7HTx4kDFjxrBs2TLA1Ax27drFqVOnUEpRvHjxAi13egVeYxBCiDvV+vXrefXVVylZsiSXL19mw4YNNG7cmLNnz7J582YAdu3aRUBAAJUqVeLs2bNYrVZat27Nn3/+yaOPPopSivDwcGrWrOmw+5AagxBC5JGBAweydetW1q5dC0Dbtm3Zvn07vXr14o8//gBMM1JERAQPPfQQBw4cYOvWrQC89NJLbN++nalTpzJmzBiH3QNIjUEIIfKMi4sLixYtol27dsyfP9+2fdiwYXTt2pWOHTuyfft2W+1h//79zJ49m2rVqlG8eHG8vLwoU6YMVqvVUbcASGAQQog85evrS3h4OIGBgTRp0gSA4sWLU6lSJbZt20aZMmVs+9aqVYuff/6ZY8eO0b17d1xdXQH46KOPHFL2VA6dx3ArZB6DEKKwmDlzJnv27GHmzJn5dg1vb2+ioqLw9vbO1XkK9TwGIYS4Uyil8rXZR2uN1WpFKZVv18iMNCWJDCIjI9m4cSMTJkxgyZIlLFy4kPfff59mzZpRp04dLBYLn332GTVq1ADMZJ3OnTsTFRWFm5sbycnJ9O7dm1OnTpGQkMDIkSPp3Llzrsvk7+9PtWrVcnzMnj17sFqtNGrUKFfXBvj+++9Zu3Yt7733Htu2baNHjx4cO3YMgIYNG9KgQQOOHDnCnj17qF+/Pkopli9fTrFixXJ8jbTfe2BgoK0dOv1nL730ksObGoRRp04dJk6cyPfff5/nQ0y11mzevJmypUrheeAA1K0LN5F2IzckMIgbSp2os3r1as6ePUvbtm0dNlknMjKSwMDAmw4MycnJeRIYGjduzMSJEwHYsWMH/v7+REVFobXGz8+Pzz//HICgoCA2btx4U3lzbpYEhcLjwQcfZPTo0YwaNYpr167l+fnLlCrF2pYtUe+/D336QPv2eX6NzEhgEJk6cOAAERERrFq1KkOyr8wm68ycOZMJEybQoUMHu8k6FSpUyPBLauzYsZw6dYojR45Qo0YNKlSowOrVq+nYsSOjR4+mb9++jBo1iho1ahAUFMT69euZN28eS5cuJSQkhF69evHKK6+QkJBAly5dePPNN5k3bx7r1q2zBaHVq1cza9Yszp8/zw8//MCCBQsAOHnyJMOGDePLL7/EYrEQEhLCkiVL6NatG0op6taty7Rp0zJ8Hz4+Ply9ehWr1cru3bvp378/27dvR2tN48aNs/0+t2zZwuuvv46bmxsDBgygR48evPDCCxw4cABPT0+7ESzZSa1NBAUF0axZMzZu3MiAAQPo378/27ZtY/jw4SQlJREaGkq/fv3o06cPR48excnJiYiICJycpAU5L4WGhhIaGpo/J09KgjfegKgo8PDIn2tkQgKDyFTaiTqp8nKyTp06dZg9ezbt2rWjffv2vPXWWwQEBDB69OgMZXFzc6Nv374EBgYSEhJCfHw8kZGRKKUIDg7m1VdfBaBMmTJ8+eWXPPfcc/z222+EhYWRnJxs9z9txYoViY6OJj4+nm3bttGyZUt27dpFUFAQY8eOJavBGLVq1WL//v1cu3aNwMBA5s+fj9baNvIkKyNGjGD58uWUKlUKq9XKqlWr8Pf3Z8aMGXz33XfMnDmTZs2aZXue9Hr06MHYsWNp27Yt/fv3Z/To0axYsQIfHx/atm3L008/zcmTJ/nxxx/RWhd4W7XIJVdXGD0aLlyAypUL7LIO+emglKqtlJqplFqslBqolPou5f1kR5RHZJR+og7k7WSd++67D4Dy5cvbXhcpUgSLxWL38MrsQX3kyBE6duxIq1at2LdvH1FRUXbnrFChQoZaTVoPPfQQa9euZcmSJXTv3p1WrVphtVp5+umns/zl3rhxY3744QeKFClCjRo1OHToEDt27MhRjQGgVKlSADg5ObFv3z4WLVpEUFAQEydO5MKFW1uK5L777sPd3d1WC9i7dy9dunQhODiYM2fOEB0dTZ8+fXjmmWcYNWqUw8fHi1tQtChUqQIFGNQdUmPIZLGeOEyQOuuI8oiM0k7UqVKlCh5pqrF5MVkn7cM/fSAoVqwYp0+fplKlSvz9998AuLq6YrGYtUpmzJjB8OHDCQoKIjAw0BY80p/H1dWVhISEDNfu3r07b775JqdPn6Zu3bpcu3aN8ePHA9CgQQN69eqV6XfSpEkT+vfvz6BBgwCwWq1ER0fbjUu/EaUU58+fp2TJklitVmrWrEnv3r0ZMmQIAElJSWzZsiXb82R23rQaNmzI4sWL8fb2JikpCScnJ5566il69+5NWFgY27dvp2nTpjd9HXF3cVhTUrrFehZpra1Kqf8ppepprX9zVLnEdakTdZ555hlmzZpl257fk3V69+5Nv379aNCgAeXKlQNMp+6bb77Jtm3bePjhhxk0aBB16tTJMkf9Aw88QN++ffnjjz/sylC5cmX++ecfWrduDcCvv/7Km2++SVJSEiEhIQAMHjyYDz74AGdnZ9tx9evX5+DBgwQEmKHgFStWJDk5OUf3NGnSJDp37oy7u7utj+Hll1+2lWHw4MEULVr0Jr6lzI0bN44uXbpgtVrx9fVl7ty5dOnSBYvFQtGiRalbt26uryHufA6f4KaUWq21fjjl9Qhgk9Z6a7p9bCu4+fv73586TFAIIUTO3MwEN0ct7RkEdMOskfob0ATTnOQCPK+1vmFDqMx8FkKIm3czgcFRfQyR2C/WM90R5RBCCJGRDGgWQghhRwKDEEIIOxIYREaXLsHDD0NQEDRrBjt2wJkzkJISIscGDwaLBU6ehIAAeOkleOcdOHUqd+X77LPMt0+eDLt35+7cqZKToVcvCAw0ZU7v++/NdxMcbO4PzPfVqpX5+/33ZltYGFy9mjdlEqKAOHxU0s2SzucC8PHH4OkJ/fubB2R8PPj43Pr5Fi40wWbAgLwpX2AgpEkwB4DVCo89BkuX3to5r16FtGmNv/0W9u+HN9+ETp1gzhxIGToLmICwYgX89Rf83//B9OkmIGzcaJ/obNUqE1TzK2WCEDkkabdF7nh5wc8/Q3S0ecj5+MDRo/DMM+bzmTPhgQdg+HDzMATz9403oHFjmDv3+rYrV2D8eJgyxQScvn3h0CHzIO7e3fzC7tfP7P/KK+b9gw/C8eNmW7NmMGgQNGgAa9eah/Hvv5tzb9hwvcx790LFiuZ1ZCR07gwdOkDr1iadQGbi4+GLL6BtW1i+3P6zn3+GlDkNBAfD9u3XP4uLM4HTxweaNjXBAcDJyRzz5JPXrxkUBKtX5+x7F6KQkMAgMurVC/z9zQMxJMT84k2VnAzz5sGWLdCtm/1xPXqYX/JffHF9m4cHjBgBw4aZB3yqWbOgXTv48cfrgWTSJPN+zBj49FOz7fx5kytm9WqzrUsXk344MtI80FP9/bdJG5Dq2jX47jt4/nlzrbSOHoXXXoNHHjH7ffstPP20/T6xsSYVAUCxYhATc/2zmJjrn4FpLgNYvNiUq0sXmDDBbCtSxNyDELcRCQwio9TEXb//bpqTPvzw+mfR0SZoODubX/Fp3XcfuLubX87ZOXgQmjc3r1P3f+89U1sYNQr+/ddsK10aypSBChXMw/pG0jeJNmxo/jZoYGooaW3fbpp8nnjC1IIyayYrXtw0f4H5mzZDbIkS1z9LW35fX/O3a1dIySUlxO1IAoPI6Ngxk+4XzEM5ba6jUqXgxAmz7bd0mUtuJslXzZrwyy/mtdVqflVHRsJPP8F//3v9QZ/2nJltS3XvvaYmkGrv3ut/q1e33/fxx2HbNvNAf/RR0/dx8KD9Ps2aQUSEef3DD6aJLJWXl2mGunIFfv0V6tQx21ODxZYt16955cr1gCHEbULSbouM9uwxzUKenqb28PnnpgkJTJ9Dnz7m136zZubzW/Hcc9C7N4SHm4fonDmm2aV1a6hXL+tjmzQxD/QhQ0wNA6B+fROwUrm6wkMPmaaiJUsynsPT0/Rt9Otn7vf0aRNcUnXubI4LDISOHcHPz+y3c6epRY0caZqyPDyuN521bm3O6+FhmtvABJWOHW/tOxLCQWRUkrh5yckmQGzbZoaOpvYHONrkydCmDVy8aJqKUtv5HSkszHS853IhdyFyq9CnxBC3uY8+gmXLIDHRvqPZ0YYONX8jIx1aDDvpO76FuA1IjUEIIe4ChX4eQ/oV3FK29VdK/eiI8gghhLiuUKzgppSqCpQEzjmiPEIIIa5z2HDVlBXcNgMRwBDgwyz2DVNK7VBK7Th3TmKHEELkJ4cFBq31Cq11c6AnUBp4D6ivlMowtk9rPUtrHaC1DihdunRBF1UIIe4qDmlKSreC2xqt9fSU7RW11mscUSYhhBBGYVnBLXV79wIvjBBCCDuSEkMIIYQdCQxCCCHsSGAQQghhRwKDEEIIOxIYhBBC2JHAIIQQwo4EBiGEEHYkMAghhLBzWweGxMRE5s2bx1dffeXoogghxG1DKVVcKTVCKdU4s89vy8Bw+fJl/ve//1G9enX69evHwoULHV0kIYQo9JRSFZRS7wPHgUlAh8z2u61WcFNKlS1Xrhz+/v7ExsYSHBzMnDlzaNeunaOLJoQQhdb+/fsBKgNHAGfgK+B9rfXuzPYv8BXclFK1gVeAUpiU2z7AvZgMq/201hcyOaYGMBToC7h3796d119/ncaNM60FCSGEAH7++Wfeffddli9fDmAFZgAfaK2PZHWcw5b2TF2kR2vdP+X9q8D3Wuu9afYJAIYDjwGJwBf33Xdf2O+//+6IIgshRKFntVpZs2YN7733Hj/99BO+vr4MGjSI8ePH79VaN8jJORwSGFIW6RkBfAwsBj4BqgDdtdaxSikXYDXQDrgIbAXKA8mAh9b6vgIvtBBCFHJKqYrAWuA/mH6E/wFztdZXlFJrtdYP5eg8jqoxACilVmutH0553QPw1Fp/kfL+Q+AkMEtrfclhhRRCiNuEUsoZ+Ab4FvhKa510S+dxQB9DENcX6fkN8Ae8gBLAEK312QItkBBCCDsOrTEIIYQofG7LeQxCCCHyjwQGIYQQdiQwCCGEsHNbB4Y0+T6ed3RZhBDidqGU8ldKfaiUCs7sc4cEBqVUNaXUXKXU4pT3n6f8+yJluFV2x6fP99E8n4sshBC3PaVUXaVUOPAP8CLQKLP9HBIYtNb/pM54TnnfT2vdD4jFTGTLlFKqllJqLibfx2vAKqCR1rpPfpdZCCFuR8popZRajZki0BX4CKiutf4gs2MKTRI9pVQtwF1rfSKTz94BngeKK6V44YUXGDJkCFWrVn0KeKqgyyqEEIWd1Wpl+fLlNG3alG3btoFJK/QW8ElmOenSKhR9DEqp+zBJ8l5Os81JKdVJKbUJky9JA/+tV68eH3/8MVWrVnVQaYUQovBKSEhgzpw51K5dm27duhEVFcX06dMB/tBaT8guKIDj+hhKKqVmAg2VUiOBDSllmaaUqpiSK2k7sBKTQ2kw4K+1Hu3iUmgqOUIIUagcP36cqlWr8txzz+Ht7c2iRYs4ePAgL7zwApgf1znikKes1vo8MCDNponp91FKLQU+BBbdar4PIYS4m1SqVIlHHnmEbt26ERISglLqls5z26XECAgI0Dt27HB0MYQQ4railNqptQ7Iyb6Foo9BCCFE4SGBQQghhB0JDEIIIexIYBBCCGFHAoMQQgg7EhiEEELYkcAghBDCjgQGIYQQdiQwCCGEsCOBQQghhB0JDEIIIexIYBBCCGFHAoMQQgg7EhiEEELYkcAghBDCjgQGIYQQdiQwCCGEsCOBQQghhJ0s13xWSpUEngWaYxaSvoBZg3ljAZRNCCGEA2RXY3gNWAt001p3A14Fyiml2ufmokqpakqpuUqpxSnvn1ZKzVZK/Z9Syjs35xZCCJE7WQYGrfVI4B/g/pRNRbXW87XW63JzUa31P1rr/mk2ddVaPwd8DXTLzbmFEELkTk76GKYC/VJev5lP5dApf48BFdN/qJQKU0rtUErtOHfuXD4VQQghBOQsMFwGzqa8vpaPZQHwB06m36i1nqW1DtBaB5QuXTqfiyCEEHe3LDufU5wDWiqlJgPWvLhoSqf2RKChUuoNYJlSagbgCbyYF9cQQghxa7INDFrrt5VS3wJKa70vLy6qtT4PDEi3eWFenFsIIUTuZBsYlFJfYvoAvJRSaK0fzf9iCSGEcJSc1BieSn2tlHotf4sjhBDC0XJSY6iTZt/a+VscIYQQjpaTzufHU/5eAz7Ox7IIIYQoBLJLiVEH+AZQKZt0FrsLIYS4A2RXY0itLaQNCOPzqSxCCCEKgSwDg9Z6nFLKBWgIeCE1BiGEuOPlpI9hOlAe2A40Bjbla4mEEEI4VE5SYlwCftdaj0eCghBC3PFyUmPYA6CUWg78m7/FEUII4Wg5CQwntdY/AgvyuzBCCCEcLydNSXWUUguVUhPTTHYTQghxh8o2MGitZwA9MQv2/JjvJRJCCOFQ2QYGpdR7mMynTsA9+V4iIYQQDpWTPoaPtdbH870kQgghCoWcNCVJUBBCiLtITjqfhRBC3EVykna7F9ApZd+lWuv5+V4qIYQQDpOTGkM1rfUTWuvHgGb5XSAhhBCOlZPO56JKqeYp+2qlVB2t9V/5XC4hhBAOktNcSW2BYOAc0D2vC6GU8ldKrVBKfaaUGpHX5xdCCJFz2S3UU0FrPS7dNielVDmt9Zk8LMe9wGqt9adKqf/Lw/MKIYS4Sdk1JfVTSlUHTgGxQJWU7TOAvAwMu4GRSqkngPD0HyqlwoAwAH9//zy8rBBCiPSU1lmvvaOUUpiAUAI4pLW+lOeFUGoo8KvWepNSarHW+obNVQEBAXrHjh15XQQhhLijKaV2aq0DcrJvdk1J5YBQoAJwEpiL6XPIa2uBsUqpp4Gj+XB+IYQQOZRdU9L7wDvAMcAfeBfok9eF0Fr/QT50aov8ceLECcLCwjh69GienlcpRb169Zg1axZFixbN03MLIXIuu8Bg0Vr/CaCU2gck53+RRGH35JNP0q5dOz744ANMS2PesFgsTJkyhVdeeYXPP/88z84rhLg52QWGeUqpRSmvrcDsfC7P3Sk+HpYuBV9faN8e8vBhmx8OHDjA0qVLKVOmTJ6f+9lnn2Xo0KE52tdisRAdHY3Vas3zcqTn6+uLu7t7vl9HiMIgy8CgtY4EIgukJHez3bthxQpwdYV69aB8eUeXKFtOTmYKTGRkJH369KFatWq4uLiwaNEiSpYsSXh4OG+//Tb79u2z269q1aoALF++nI8//pjp06fz7LPPMmHCBLvzZufQoUO0adOG+Ph4nJ2d8+EOr9Nac/XqVRYsWECXLl3y9VpCFAY5mflso5R6T2v9en4V5q5VtSqUKwelSkHJko4uzU3r1asXEyZMYP78+Xz55ZcMGjSI1atX07hxY/bt20ft2rXt9ksVGhpK8+bNiYiIuOlrvvXWWzz//PO8+eabeXYfWdmyZQuPP/64BAZxV8jy55lSKkIp9XXKv2+ArgVUrruLnx9MmQIjR8Jt3FwRGxsLQFxcHBaLhbCwMJYuXXrD/cuWLXvLfRTnz5/n/vvvt72PiIigVatWtGzZkq5du9K5c2eaNm1KUFAQb7zxBkePHuWZZ57JcJ7w8HBb4AJTs6lcuTLBwcG0bduW8+fPA3D//ffbXgtxp8uuxnBUa90/9Y1SakY+l+fulcMHZExMDJMmTeLMmbycX5g9FxcXOnTowOOPP57hs/DwcFauXInVamXz5s2sXbuWjh070qJFCyZPnmy33+bNm6latWqedC6nBpXo6GjGjx/PqlWr8PHx4eDBg7z00kssWLCAGjVqANxwBFVWNZu0NaC87GQXorDLLjC8mvaN1npgPpZF5MAjjzxC9erVadOmTYE+rK5du8aQIUPw8PDI8FmvXr0YN24coaGhHD9+nBUrVnD8+HG++uorDhw4wIkTJ2z7pW1KyiurV6+mV69e+Pj4AHDvvffi5+eX7XHpazZpaw5wvQYkxN0mu87nS0qp94HU6dFXgIVa60P5XjKRQXJyMlu3buWHH37I9w7XzMTGxvLjjz9m+pmzszMjRoxg7NixXL16le+//x6AtWvXsmzZMurWrZtv5Tp9+nSm5+/Zsyeenp4MHDiQpk2bZvg8q5pN2hqQEHebnAwBcQK+Br4CSgJv5GuJxA1prVFK2YJCant4mzZtCAoK4ssvvwRg586dBAcHExQURL9+/YiLiwPML+SwsDCCgoJo0aIFX3zxhV3b+9atWwkODiYuLo6aNWsSFBREUFAQf/1lsqy7urpisVhuWL6aNWuyaNEiypYta9sWGBjIypUrM91/7ty5DBkyhAULFvDiiy/e8vfi5+fHv//+m2H7ggULiIyM5Iknnsj0uBUrVhAeHk6HDh3Yu3evXc1m165dBAQEcPy4rGwr7j45GZVURGu9A0Ap9RyQ8f9A4TCpzTPx8fE8/vjjVK9encGDB7Ns2TLKlCnDl19+yfjx43nnnXcYN24crVq1YtasWWit+emnn2znOXjwIEOGDGHFihV4eXlRunRpIiMjs71+avBIlT73VpEiRVi/fr1t37T69+9P//79ya2OHTvSvXt3nnzySXx8fDh06BCnT5/O8pikpCQuXLhww5pN2hpQasAV4m6RkxpDhFLqm5RRSd8D3+ZzmcQt8PT0ZMiQIXz33XcEBwfbJp899dRT/Pzzz4CpEfTs2RMwHbctW7YE4OzZs/Tu3Zv58+dTunRpAC5cuEDLli15/vnnuXbtmt21PDw8iI6Ozpf7iI6OxtPTM+MHyckQE5PpMaVLl+att96iU6dOtGzZkqFDh+Lm5pZhv4iICEJCQggJCWHhwoXUr1/f9llmNZuaNWty7ty5Au/oF8LRclJj+Ba4mvJ6ndZa0mIUUuXLl2fevHkZZg5nN2ns119/pWvXrlSvXt22bfPmzfj6+vL2228za9YsXn75ZdtnL730Eh06dKB58+Z5nhLjhx9+4KOPPrL/QGt4/33Yvx/CwqBFiwzHpj7wb6RKlSpZ1iJuVLPZuHEjAAkJCTdxJ0Lc3nISGD7l+uznT4Hc1/1Fvjh16hR9+/bN0N6eXcqIzp07Y7VamT17Ns899xxgUkAAdO3alSlTptjtP3z4cFq0aJF9Er1z52DrVqheHe67L9vyK6UYPnw4DRo0sP/AYoFjx+DaNTh50rY5u5TxeakgryWEo+UkMFzSWocDKKXqZ7ezcIxr167x4YcfMnLkSAYPHkxUVJStjyF1RE7z5s1ZsGABPXv2RGvNli1bqFixIgCzZ8+mXbt21KxZkwceeACtNe7u7mzZssWuJpEqMDCQwMDArAs1ezY4O0NUFPTsees5oFxc4PXX4fBhaN4cgJIlS7Jr1y7at29/a+e8Sbt27aLkbTgrXYhbkZPAEKOUmpvy+lh+FkbcvPDwcH7++WfbePymTZsybdo0evTogdaaKlWq8MknnwAwZswYBg8ezOzZs0lMTOT555+3BQZ3d3e+/vprOnTowIwZMxg4cCBFihShRIkSzJ8//9YKFxwMR45As2a5TwxYrZr5l2L8+PGEhIQwZcoUXFxuKrPLTdNac+XKFRYsWJCv1xGisMh2BTcApZQPoLXWV/K/SFm7m1dwS0pKwsvLi6SkJIdcf8qUKRw/fjxD05KjWCwWzp07VyDZVUuWLCnZVcVtLS9XcPuG65PblFJKa6175LaA4tY4OTmhtSYpKQlXV9eCvfiRIyREReX7r/Ob4ezsTLly5RxdDCHuONnNfLYlxlFKlddayxwGB3J2diYkJIQnnniCVq1aFVxKjKgorq1dy9S//uL/ZAEdIe54N/PzbyjwWn4V5E5z7do1Pvnkk0xn5OZGzZo12bVrF1988UWGz1xcXKhZsyYlSpTI02sSG4trfDxftGpF21at8vbcQohC52YCw+58K8UdqGfPnsTHxxMcHJznv+wrVaqU6faoqCjCw8PZsWMHFSpUyNNrcuSIWUhImm6EuOPlODCkDlnND0opJ+C/QFFgh9Y648/h28yqVauIiYnBy8urQK+7f/9+24Q1wIz9X78eypSBBx649ROnrMjknNAAACAASURBVLwmhLjzFZaexEeACsAF4GQ2+94WkpKSbCmq0y5rmZSUxJw5c6hcuTKDBw/m4MGDJCUlERYWRp8+fYDrs3hHjBgBQN++fdm3bx+urq60adOGcePGkZSURMuWLfn999/Zs2ePbd0BDw8P+1FLv/4K8+eDlxfUqGFWibsFhw8f5rfffiuQEUDp+fn50axZM1kTQYgCUlgCQ03gZ631p0qpxYDdWo9KqTAgDMDf398Bxcu91GR3W7duZebMmXh4eGSa0O7ChQv4+vqyadMmW2AAbIvOtGvXjsuXL1OkSBGWLVvG8OHDs75wpUpQogSULQsp6xXcrE2bNvHYY4/RvHnzAh+VpLVmz549PPHEE0yaNKlAry3E3aqwBIaTQGLK6wx5nbXWs4BZYOYxFGC58tylS5coWrQokZGRvPvuu4B9Qrvly5fz+OOPs379ek6dOmXXV2CxWLh69SpWqxWllF166xuqWhWmTTOzh29xDYcZM2YwYcIEnn/++Vs6PrfOnTtHpUqVePvtt6XWIEQBKCyB4VvgI6XUg8AmRxcmP4SHh7Np0yb+/vtv1q9ff8OU1uvWrWPu3LkULVqUZcuW2dYp6NmzJ8eOHSM0NJRixYrd3MVzOTErLi6OcuXKcfHiRR555BEAdu/eTcOGDalatSp169Zl2bJlWK1W7r//fqZOnUrNmjVtq6gNHDiQ+Ph45s2bB8DevXv54YcfiI2NZePGjXarunXu3JnY2Fjc3Nz44osvqFixIqVLlyYpKQmr1eqQBYqEuNsUisCgtY7jDk/Ol9qUdPbsWUJDQzPd58qVK/zyyy889thjWCwW3NzcbIFhwYIFeHt7Exoaaluwp6AVK1bMFtACAwOJjIzku+++Y9myZfz4448opWwrvGW2nkPfvn1JTk4mICCA+vXrZ7oa3LRp06hatSobNmxgypQpfPDBB/l9W0KIdHKyHoPIQz4+Ply6dMmW0A5MO/rmzZtZs2YNb731FmvXrmXDhg34+vpy4cIF27F+fn7Uq1ePdevWOar4GXz99dcMGzbMFqhaZTPPYdOmTbRs2fKGga1qyugnFxcXqR0I4SASGApIeHg4QUFBtG7dmmHDhjFmzBh+/PFH2zKbhw8fZunSpXZrAbRo0SLD4jGhoaHMmDEDgB49erB+/Xr69OnD8uXLC/J2bE6fPm1rMkrr3LlzttXdDhw4YNv+7bffXh9KewMWi4WJEyc6rE9DiLtdoWhKutMFBQVx7FjGxLSzZs2ye586XDXVgAEDMmyvXr26LQh8/fXXeV3Um5a63vI999xjtz2zpqTUmtHUqVOzPOeQIUPo3bt3pum+hRD5T2oMeUFr8y8NZ2dnh2RBTUpKyn0TzC+/wKRJ8Pff2e7ao0cPJk+ebFvIZtOmG48d2L59O40aNcqyfHPnzkUpRe/evW++3EKIPCE1htyKioJ33oGiRWH4cEhZr/jBBx9k4MCBhISEFFhHcVRUFFu2bMl9h+2iRXDmjEmBkW6Z0PQ6dOjAvn37aNWqlW1UUurQ2/SWLl1Kt27d7LYtWLCAX375BYBRo0bxwgsv0KRJE4KCgmjVqhXjxo3L3b0IIW5ajtZjKEwK3XoM27fD//5n5glMmACVKwMQGxvLf//73zxPopcVT09PnnvuOZo1a5a7E23YYNJo9OwJDRrwyCOP8Oyzz9qGqjqCs7MziYmJ0iEtxC3Ks/UYRA7UrQuPPmpmFadJble8ePHbd6hl27bmXxqOSIWRSmstay4LUYAkMOSWhwc88YSjS5GvatSowaJFi6hXr55DUmJs3LiRypUrS21BiAIigUFk67///S+hoaGEhIQ4LImeo4bjCnE3ksBwl0tMTGTZsmVER0dnuV9gYCCBgYE3fX4nJycaNGjAA7lJ+S2EKFASGPLDkSOmAzc4GNKN7y9snnrqKU6fPk39+vXzZfRUcnIy48ePZ+rUqTz++OPZHyCEcDgJDPlhwQL44w84dgwmTnR0aW4oNjaW9evXc/78edzc3PLtOkuWLOGzzz6TwCDEbUImuOWHBx4Ab29o0cLRJclSQkICXl5etqAQGRmJi4sLUVFRgJmQppSiYsWKxMXFATB16lSmT59O3759OXTokO1cqc1MGzZs4IEHHiA4OJj9+/cDJvleQkJCQd6aECIXJDDkh5AQmDMHOnZ0dEluWoMGDWwdvUuXLiUgIIAXX3yRyZMnc/HiRRYvXpxlDqPx48cTERHBwoULGTNmTEEVWwiRhyQw5JfM2us3bzZNSzlINZElreHCBUhOzt15MtG6dWsiIswCen/++Sf/+c9/eOqpp9i0aRODBw9m+PDh2Q5Z9fb2xs/Pj8OHD+d5+YQQ+U/6GPJbXBycP2+W1nzvPYiPN+svv/rqrZ/z++/hiy+gfn0YMiTvygq4ubnh4eHBL7/8Qu3atTlz5gxg0lWMGjWKTp062fbt2bMnnikpQNI6e/YsMTEx7Nu3L0/LJoQoGBIY8tPx49Cjh3n95JNw8aIJEg0b5ux4reHgQfD1hdKlr28/eRKuXYOjR80+eTyaqGPHjgwYMIBZs2bxySefAFClShWqVKlit1/qOtRwvY/hvffe48knn6Ry5cq0KOR9LEKIzElgyE+//WYCQWqW1Vq1oEQJaNIkZ8dv3QozZphj3nvPlqCPbt2gfHmoWTPPgwKYwLBu3ToaN25808c2a9aMH374gb///puPP/44z8smhMh/EhjyU/Pm0LevaToaOBAsFpOx1NU1d+f18cmQy+iGLl82mVKrVYMcppQoUqQIc+fOvaWiTZw4kY0bN1KyZEk+/fTTWzqHEMKxJLuqI/z7L2zZYoa1urvDZ5+ZrKxPPgnR0XD1qi1La6ZNSZcvmxxN2QUYrWHUKDOfokcP6NLF7uOzZ89Sr149zp49m8c3aG/jxo288847bNy4MV+vI4S4sdsyu6pSyhvYBIzRWq9ydHluidams9nbO+v9vvgCdu+G3383tYqdO+HPP02geP99ExiGDYP77jPNRWn9+Sd88AFUqACjR2cfHBITwWq93pyVhlIKi8Vykzd58ywWS4GtSSGEyL1CExiA4YDj16rMjcmTzfyFoCD45JMbN93UqQP793OqfHl2RkWR7OJiagWbNpmhrImJsG6dqS2kt3cv7N9vOp6//vp6v0MKZ2dnGjVsSCV/f9P/8MYbcOKEuWY6JUuWxM3NjenTpxMQEJDx4Z2UZNJ7lCtnFiK6BcnJycycOZO6deve0vFCiIJXKJqSlFIhQCnAA4hOX2NQSoUBYQD+/v73Z7Z+cqHw2GPw66/g52ce7CVKXP/MYgEnJ1tn8Z8//0yb9u0J8PPD7d57r//yj4838xN8fDK/RmIinD0LxYpl+rBOOnaMX//6i7WzZtGwV69si/zH77/z+gsvEH35MqRLi5Fw9iznz53D6ux84/JkQymFn58fbdu2veWU3a6urrRv356mTZve0vFCiNuzKSkY8AbqAPFKqTVaa1t+Z631LGAWmD4GxxQxG3v3mn6AZs3gwQdh5EgICIB+/cxnH31kRiOFhYFSzPvsM57z9+e/991nls8MyMF/r+Rk02fg6wvPPWeuk96LLzIpKYk5s2czPQeB4b7ERNZUrmxqHu++a84NREdH06h2bbpUrIhf+fLQps3NfiN5Jj4+ns6dO7Ny5UoJDkIUgEIRGLTWIwGUUn0xNQbHLRd2szZtMplUFywws5ErVTKdvBcuwE8/Qa9epi/h6lVTmwgNBWdn4lxcqHL//Vxs2JBHXnsNnJzYvXs3DRs2pGrVqvTs2ZOQkBAAjh49SrVq1Tjw++/cc+kSY3ftInDTJupUr06nTp3466+/uHLlivlFHhZGuStXOHjtWs7K7+FhliV1dzd/U2zfvp2a9eubeQzly0ORIvnx7eWYk5MT3333nQQGIQpAoUqJobWed1t1PCckmBFFixdDbKzpeE5IgNq1oVUr6N/fNBF17GgCRunSZpIbmGalJk0oNngwkZs2ERkZSd26dYmMjKRy6oikNGrVqsXUGTNMTaRZM7j/fnx9fYmIiLBf66B+fdPH4eWVs3v4z3/MWtXjx9s1TSUlJeHp5QX33gtFihAREUGrVq1o2bIlSinq1q1LrVq1qFmzJkFBQXz11VdUqVLFNsw1MjKSUaNGAaY5KTXNxrx585gzZw5Hjx6lbNmyBAUF0bRpU3766ScAPvvsM6pWrcozzzxjV0xPT0+SMulAF0LkvUJRY7htubmZABAbazp3LRa4/36oWtU8+EuWNPvFx8Pq1aYz99574dlnzXatzQxmD49sL9WoUSMOHTpErLe3GZHk4oKHhwceOTg2WxUrZvlxdHQ048ePZ9WqVfj4+HDw4EF8fHxYt24dycnJhIaGAjB58mTmz59P//797Y6vXr0606dPp0265qi2bdsyf/58/v33XwYOHMiDDz5Ily5daNmyJWPHjs39fQkhbokEhtxQyvQh9OtnHvAnT5rmo549TSdxjRpmaOmFC+bhn5xsag0REaYD+dQp2LULXnnF/NLPRmhoKLNnzy6AG7O3evVqevXqhU9KB/S9996b6X7u7u60aNGCDRs24JpmGG25cuUoUqQIBzMbZQVcunSJoim1lVKlSnHlypU8vgMhxM0oVE1Jt6UTJ8y/6dPNvIKvvjIB4PJl06FrscD8+SaJ3muvmcAwZ45Jl3HypOl7yGEW0q5du7Jy5UqS4+LgwAFTUykAp0+fxs/PL0f7vvjii7b8Smm9/PLLTJ061W7bhg0bCAoKIjg4mN69e+dJWYUQuSc1htw4cQJSmzyKFTMTyQIDoXp1kxepenXToRsbaz6vUeP6nABvb9MX0Lx5xvQWFy/C55/DuXNmNnQKZ2dnunTpwmcffEBQzZrmvCNH5vtt+vn58e+//+Z4Xx8fH9siPakCAgIYNWoUNWrUsNU8UpuSrl69ysMPP0zbnKb5EELkK6kx5JTW5oFttdpvS/3XtatJO/Hcc/DUUyaDatGiphP4jTdg8GDTH1GlihkWWqmSqVF065ZxjsCpU7y+Zg0hr71GSOvWxMTEmO1//03o4cMcj44GJyeSPD0JCQlh7969tG/fnm3btuXLrXfs2JH58+dz+fJlAA4dOsTp06dvuP/LL7/MtGnTMmy/UVOYp6enbYU4IYTjSY0hp1asgCVLIDjY9CmA+cVetqzpH5g6FXbsMDWGGTOuJ8tzdjadu6dPm/4Ef3/TDPTXX6ZGMHSoqU0AmzdvBmDstGmM/b//M6ObgoLAx4f58+fDwoUUP3OGK4GBpqYQGMjG117L91svXbo0b731Fp06dUJrja+vb5ZJ9gICAvBNmQ+RVteuXRk6dKjtfWpT0rVr1wgLCwNg1apVvPPOOxw+fJjHHnuMJUuW5P0NCSGyJIEhp/75x4wuOnTINCEdOwbLl5u/58+b1BFRUbBypakNpC7O88YbZnLb8uWmE7pxY9M57eqK1d3d5Em6/35bcABMc9Prr5u+iBdfhGeegXbtoHVrE6C0Nn9vMOnMas37aSAhISG2eRWp+vbta/c+NbDd6LWzszNHjx61bc8seV+nTp3sFgMSQhQ8CQw51asX3HOPqR28+67pOD5/3owsKlPmehoLb2/Ys8e8Tkw0r9esMbmNvL3NQ93fnxq9erF01iweev993Jo1gz59Ml5z2zYTTLZtu55Mr1Mn+PZbKFUKjh3j0IEDDHrlFY4fP247LCEhAWdnZxYvXnzLt2uxWEhMTLT1B6RSSnHffffxzTffUKFChVs+/81KTk7GyUlaPoUoCIUiV9LNcFjabavVjCBav97UAFLt2mW2KwXFi5v5DLVqmf6Ee+4xw1SfesrMYRg0CJo2hSZNSLZYeLl1a9bt3k2yh4fZL/2DLznZDIN1dTUzk1OlrtqWnMyZM2fwdnXF290dlZyMs5sbLTt14n8ffohbutxHN+P48eO0bNmSkSNH2o1IslqtLF++nOPHjzNixIhbPv/NiI+P54033mD69Ok8/PDDBXJNIe40N5MrSQJDTn36qZl/8MsvZmJb166m1rBmjXl4W62mM7loUTP6aMwY0x9x7JhJoX3qlGmKcneHli0hJsb0HzRpYlJmfPsthITY1xyio2HECHP+oUOhQQP7MkVGUr5DB7b7+lLB25vIpCT6nDtHtUaNcHF3Z9GiRZQsWZLw8HDefvtt2xrMkZGR9OnTh6pVq5KUlMScOXOoXbs2EydOZPr06Tz77LNMmDCBLVu2MHHiRFundr169ShWrBgxMTHs3buXoKCgAvnqXV1defTRR+nZs2eBXE+IO9HtmESvcPv7b/jmG9NhrLXJG7RvH/z8swkI//mPaeo5etS8r1MH6tUzx65eDVeumP6HqChz7OrV5mG/b5/55X/6tHmffgKYi4sJQhZL5rOjmzQBd3dU8eIm8JQoQa9HH2XC/PnMnz+fL7/8kkGDBrF69WoaN27Mvn37qF27NgC9evViwoQJbN26lZkzZzJ16lRCQ0Np3ry5LX1FixYtcHd3Z8+ePTg5OfHCCy+wbNkydu7cSVhYGMuWLcvHL10I4SgSGLJz8aLpUzhyxORCcnExfQXff29qAGCafE6dMg/4Bg3Mr/y06TC8vEwq7vLlTS6lIkXMaCUXF5N59emnTS3Dw8PUMCpVMvmXLlyA4cPNfpm153t5mVpKw4am9nL5sqllALEpk9/i4uKwWCyEhYWxdOlSW2BIlXbWcdmyZW21ilQXLlygUqVKKV/Fxbz6VoUQhZgEhuy4uZn0FfHx5vXly2bWctqEbn/8YfoHnJ1J+M9/0KdPm6yrVis88YSpUfj7m+U8V6wwfQZ9+phmqEOHzN8WLWD6dJS3N+6vvmr2S0oynd3Nmt24fErBm2+a5UKPHyd8715W1q+P1Wpl8+bNrF27lo4dO9KiRQsmT55sOyw8PJxNmzbx999/s379+huePu0Ip/wY7SSEKHwkMGTH0xOqVTMP+e3bTZNPelrzr8VCB4uFv777DueNG00tAkzAsFrNAzy1P8fJyQQDq9U0E61YYetMtgA1t25lzRNP4G+1mg7srMTHw9tvm9XjoqPptWkT48aNIzQ0lOPHj7NixQqOHz/OV199xYEDBzhx4gRwvSnp7NmzhIaGsnLlykxPn3YkkIwKEuLuIIEhK4mJ8L//mY5iHx9TW7iBkcBDTk7sqV8fNWiQaRJKTjYjlbZvN7/+//zTDDP97rvr8xYsFtOsdOkSzJqFrlqVMdu2MWLJEhY+8USGpTvtWK2maeryZdMkVbUqYOYLjBgxgrFjx3L16lW+//57ANauXcuyZcvsltn08fHh0qVLN7yEr68vJ0+exMnJiWJp51oIIe5Y8hMwKzExpsP5yhXzKz8hIfP9lOKskxMtrVbUX3/BunVEuLjQavNmWi5dSteTJ+n87780jY4m8NAhxowbB5gFeJ5JGYW0de5cgt97j/glS2hZrhxno6JMc9TcuWY4LJjO6cjI6+VwcjL9HbVqZZjsVjM5mUWLFlG2TBnbtsDAQFvNIDw8nKCgIFq3bs2wYcMAmDt3LkOGDGHBggW8+OKLAIwbN44nn3ySxx9/nHEp5RZC3NmkxpCVxETTB7Bvn/nl7ux8vYkoLa3B2RmlNTg5Eb18OeMXL2aVvz8+PXpw8J9/eOmPP1hQuzY1nn6adsuX2/IOARw8eJAh4eGsePRRvOrUQd1zj1kvOiEBNm82D/4WLeD99+HSJRKvXuXoPfdgsVhISkjg7507uThjBmV796Znz57si4iAjz7ir86d4dFH7TqUUzOcrl271u4W9u3bR/PmzWnevLndtooVK9rNYhZC3PkkMGRl6FAzbyE1gV5Wcz4sFjMSqXp1Vm/fTi+rFZ/jx2HRIu4NCMDv9GmIjcWyaBFXrVasSUlw9SpnT56kd+/eLPjmG0pXr27OtX69CQytW5tO5WrVTMd3mTKcuniRoIEDsbq64ubmRszVq4Ru24bLb79B6vBRi8XkYdLa1HjSrI1wM7TWnD59mk8++UTmEAhxF5HAkJU0zTBZBoVUSkF0NKe1pq5SppZxzz2mSSg5mZ4xMRzbtYvQQYMo9v77xKxaxa/799O1XTuqpwYFMLWEI0dMDeXjj68/2EePZtrrr9OhQwemffABFCnCvffcwxdvvkmzbt3s8y3Fxl4PVrmwe/du2rdvbxcYjh07RhEHrwEthMg/EhiyMmmS+RV+8uSN93F2Nr/mExJM09OFC/g5OfGvu7tpAvr7b9NHYLWyoH59vJs1I3THDnTZsnDpEp1LlcIaF8fs2bN5LjTULNqTWjtJTYeRyt2d2JMnaXT6NIwYQWS7dlyIiaFlWBjer75K3bp1OXToEPHx8SilqF69Oj4+Ply7do39+/ejtUZrTZ06dYiPjycmJoZq1arZTn/w4EHOnTtH1apVKV++PGCGqEZHRxMcHAyYHEp//vknCxYsyJevXAjheIUiMCilHgUeBsoA07XWNx5YXxC0hl9/hdmzsw4KcH12cny8+YUeEEDHhAS6//47T549i4+zM4cqVOD0pUtQsiR+ycnUs1hYB9QKC4MffmD2kiW069KFmnFxtNy+Hc6cMbWV9Cm1DxwwuZni4sy6DvHxDBgwgGeffZZjx44xa9YsWrVqRZs2bYiJiWH06NG8/fbbzJkzh+7du9OoUSMSExMB+Ouvv9i5c6fd+sznz59n+/btWCwWW06ipKQkHn74YUaPHg2YJHo1atSgYjbrRAshbl+FIjBorZcBy5RSJYDJgGMDw+HDZqnOnTuz3zch4foooZTmo9K+vrxVsiSdtmxBe3jgW6cObkpBWBgkJxN6+DCvbdrEVD8/+Pdf3K9c4euvv6ZDYCBL/P3N6KPkZBMA0q5rcPmyqaGUKwfPP28C2G+/Ua10aapVqcJbb73Fl19+adt99+7deHh4UKtWLWJjY7n//vtts5zd3d05f/68rSaQ6sqVKyQnJ9u2JyYmopTKsJ8Q4s5V2IarjgKmp9+olApTSu1QSu04d+5c/pZg82azdvORI2YNhpuRmGg6jvfuJeSee/jx/ffZtGwZy9asYeXq1dRo1w58fKheowbLhw+nyvHjzNcaxo2jbLFi7Nq/n6rDhpl+CScnUxtZvdqsChcZCY0amRxM3bub7K1A+Jw5BN1zD/0CA1FK2RWnYsWKnD59mmHDhhEXF0fjxo15/PHHuZo6/FUIITJRKGoMyjzR3gG+01rvSv+51noWMAtMdtV8Lcy+fWYhnsOHM5/lnOIU8CmQOuh0HzArKYkNVquZ3LZsmVnxzcvLzDE4fx42bjRzIzw9zboKUVEmH1JkJLzwgq3z+LiHB38Dr777rglUly7hcfAgT330EZQubdcp3qtRIyb4+EDFigT++qtdGU+ePEnt2rXx8fHhgw8+4IMPPuCdd94hPDycWrVq5eW3JoS4gxSWGsNLQAjQXSk1wKElqVrVdBhHRd1wlzigFRAN+Kf8KwFoZ2f8XVzwV8r8s1jwr10b/ypV8D91Cv/Ll/FPTMQ/Lg7/a9fwf/hh/IOD8X/gAbOPkxP+5cujPTwoXrIk/v7++LdsiX+JEjiXKUPbtm3t5j8AJpNrr17wyis0a9bM1pQUFRVFREQETZs25fDhw6SmVy9durTkPBJCZKlQ1Bi01tOAjKvHO0L58uaXeybLTqY6CHgCn4AZNZSURFugbVIS+52czJd65oxJo/Hbb/DTT6Z/IDHRNA85OcGPP5p1GFxS/hOsXm1yJRUpQkxyMuvWraNBgwZm+8mT4OnJocBAzpw5Y18YT0/o3BmAsWPH8sILLzBz5kyUUkydOhV3d3c2btzI3Llz8fb2xsfHhwULFrBz504WLFjAL7/8AsCoUaPYsmULCxcuRGvNv//+a+twFkLcXWShnvRiYsway2vW3HCXHcCAlL8XXVx4JGU29C6gJlAB2OfsTFE3N3BxYZKrK37JyZxOSqJ9fDwbmzWjTLFiUK8eTT/6iP/Uro0+fZqGXl68MHgwNfr1Y9iwYXzzzTdMeuUVQo8dg2rV6LNvH8dPnODJJ5/k+eefz7/vII3ExESKFCliG80khLg9yUI9uTFpEqRLF5GVYsnJRBYtCpcuEQhsBsYCg5ydCfH2NiOVLl8GLy82ak0/Hx/2FytGsJ8fFCtGAycnNpcoAWvWMG7SJH5ISqK+tzdvvfUWTZo0ITk52azJ4OwM/foBkG0wj4oytZWsEvDl0O32w0EIkXuFpY+hcIiJMc0+N9sGn1l20qQks91iMfMcAgP5oWZNPvjoIzakrgM9b56Z//DPP3DmDG+89x5rvvsOwG6dZVxdbetB+/j4sHnz5hv3E+zYAa+/DuPG2a8ZcYt++uknyqSdAS6EuONJjSGtCROw/vILGkg78DP1depv57SP5OSUf84p76+6uaETExmiNSWSkuDCBda7unLx998p+cADFAsIoMicOcR++CHF4+NNTaBzZ7j3Xtzc3EjK5mH+0EMPsWjRIjw8PHBxyeQ/n9V6PdHftNx122it8fT05JtvvsnVeYQQtxcJDKm0JnHJEo4A3pgmoWJAEOAOJAH/AjHAFuAo8BngAUwB7gMSgG1KEeXmxvsVK9KuXj30wYPEHTnCR3Fx7PrzTx566CGirlxhVb16PNOokRke++GHoBSJiYm4ZpPwztvbm02bNpGQkIDFYsm4Q2Ii7N5thrSmrM+QG56enhnmRwgh7mwSGFKsXbeOD+PjOYrpQP4eqAQMAR4F9gD/B6SucxYLzAQaAr8DZ4BEoGdSEheADVYrb3TqRMfmzfnfkCF8uWcP/zdlCkGHDxP344+Enj3LM8uWwYMPsnPXLhYvXsz58+fp1KlTjsrr7u6e+QdeXiCzlIUQuSCBIcW0adNYlzJ34UDKtn1AKDAGuADEpzvmD+BPTE0hNaOSi5MTycnJHD56lAEDBtCoUSOaNGmC0x9/cO3aNUK3biUoOJiz8+cz7NbMrAAAC3tJREFUa/Zsdu3aRffu3YmLi6N69epMn24mfk+cOFGGjgohHEICAxATE8PGjRtv+PmpG2xPHygAXF1d8fb25sqVK5QvX55Zs2ZRp04dgoODWbNmDVfj4tjy6698+OGHvPPOO3Tt2pUnnniCPXv20KFDB1tT0siRIxk5cmTub04IIW6SBAbgxIkT2Xb6ZkUpZRvWWbx4cby8vEhOTqZ48eLUqlWLv/76i8WLF5OYmEjjxo3p2LEjlStXplatWri6utKmTRu6dOmS7XWSkpJwdnbOdj8hhMiNuz4waK3Ztm1brs8Bpt3/ypUrnD9/Hq01J06cYOHChVy7do2YmBhcXV3x9fVl79697N2717Y4z9KlS7O9xvnz59m4cSNjxozJVVmFECI7d31g+OOPPxg4cGCuzuHi4oLFYiE5OZnk5GScnJxwd3enfPnyREREkJCQwNmzZ/Hy8iIyMvKWRvl4enry7bffUrNmzVyVVQghsnNXBwatNUOHDs182OdNnsfNzQ2r1Yq7uztlypRh5syZBAUF4erqSmRkJNOnT8fDw4PJkydTtmzZPLoDIYTIe3d1YIiOjmb9+tyvCWS1WnFyckJrTWJiIomJiVSoUAFXV1d+/fVXli5diq+vL02aNKFUqVJ5UHIhhMg/d3VKjClTptzysUopXFxccHZ2plixYpQoUYJixYrh7OxMQkICERERAKxcuZKzZ89SunRp+vfvL53HQohC766uMUyaNOmmj3F2dsbJyYmkpCQsFottopnFYsHPz49ixYrh5uZG+/btAejWrRvLly/nkUceydOyCyFEfrlrA8PXX399S8elNgVduXIFNzc3ihQpgp+fH//f3tkHWVnVcfzzLSBhWQlzK1xqVyZrBiHacjbEGaAXBZWScApKcMBMMsC1BGtLJprUKTFGgRTZRaJBhAKHZJDRoJyIaWxklsx4qelFXnqxQpaXQHD21x/nXLh32+fuS7v73Lv8PjM7997nOec539959p7fOeee33l69erFkCFDqK2tpaKigpKSEgCqqqqoqqrqNN2O4zhdzXnpGBobG5k8eXKH8tbV1VFdXc3GjRvZt28fO3fupKysjJqaGqqqqujfv38nq3Ucx+lezkvH8JEO7CVUWVlJTU0NEyZMQBIzZ87k9OnT1NXVcezYMYYPH+5OwXGcHsF55xj27NlDQ0NDm9P37t2bGTNmMH/+fMrLy3NiEPr06cOsWbO6QqbjOE5qnHeOYejQoW1KV1FRwaBBgygtLWXMmDEMHjy4i5U5juMUBgXhGCSVAI8Qdq5+3sye6KJyWk0zevRotm7dSlNTEw0NDezdu5fx48d3hRzHcZyCpCAcAzAJWG9mmyStA7rEMSTRr18/qqurmT59OlOnTj0bazBy5EhGjhzZnVIcx3FSR4XwsHdJtcAWM9slaY2Zfa7Z+duA2+LH93HukQkd4WLgX/9H/kKg2G1w/elS7Pqh+G1IQ3+FmZW1JWGhjBgOAoMJD0r7n2hsM1sOLO+MgiS9aGZXdMa10qLYbXD96VLs+qH4bSh0/YXiGJ4Clkq6nnNPz3Qcx3FSoCAcg5mdAGakrcNxHMc5PzfR65QpqZQpdhtcf7oUu34ofhsKWn9B/PjsOI7jFA7n44jBcRzHyUOPdgySSiStklQn6aas48MkPRH/hqWpsTXy2HCdpGckzU5TX2vk0f+1eGyTpIINK8+jf6KkZZI2SyrYYJck/fHccEmvSiroTb7y3IMFktbF+3BJmhrzkUf/IEmLJS2RdFWaGpvTox0D5wLnvgB8Mut4DTAL+BIwJw1h7aBFG8zsGeCB1FS1nST934nHHgfav6th95Gkf6OZfRG4B/hwWuLaQIv6JfUGbgW2pCWsHSR9j98g7JZwBjiShrA2kqR/LnAMaCIs2S8YerpjGAwciO+zH+w8wMyOmFkjUNr9stpFkg3FQqL+2FP9DLCxu0W1g3z65wH1wLbuFtUOkvTPBZYAxfAjY5IN95vZNOCnBCdXqCTpvxz4IbAAmN/NmvLS0x1DJnAOcm1tlDRA0oUEj13IJNlQLLSoP9b9o8DdZlbI9yCx/s1sIXAtcFd3i2oHSfo/AMwGqoGZ3S2qnbRog5k1xbevAoU8HZZ0Dw4CrwHHgQu6W1Q+evSqpLg531LgFPBLYLyZTYu/K8wDBDxgZi+nKDMveWy4EqgFBgIPmdmGFGUmkkf/U0Bv4BDwIzP7WYoyE8mj/1ZgBDAAqDOz7SnKTCRJf9b5HwCzzex4OgpbJ889+DrwLsL2EneY2d9SlJlIHv1DgbsJo7Z6M9uRoswcerRjcBzHcdpPMU5NOI7jOF2IOwbHcRwnB3cMjuM4Tg7uGBzHcZwc3DE4iUiaLmlXfN9b0iuSJnTwWldImixpbFuitWNUa7ui0iUtkrRQ0sD4uVLSCUnl8fMaSQ92RH+zcj4aI1kfiWVWStoVI3CXxUjXLZKWSvqJpBEx3x+z0rxb0pSoaamkG7KuL0mPZf1d3A5tid/pqLNB0sOSVkq6Oh5fGF8z9fchSU9L+nRHynGKn4LYdtspaPZJGgW8A/gVgKTLgZvisXqgAVgB/Jmwrn8MsBZ4HhgGLCKsMy8DLgNGSToCvIcQEfqypLVmNkVSHWFd+ihgfWy8rgf6AhvM7LmooQ/wGNAIHCY802M8sNrMXsvSvxmYKmklIVIWSb2A+4C3AG8G7gRuB4YATWY2Ny7j/ANQDmzMKleEyPmJZmZRxyXAVjObmylU0gkzmx2XFY8FfgM0xGjpTJpFwC1mdrRZnQ8kBGFOyUo7DphMiPD9LjAReD9wYdQ/B7gIaJB0sKU6i2yL9gl4WtJW4NLohMcDq4HPEtqGQ5JuB94bNd0D3Av8BfhtrMcro4bFUVNprOe9ZrYyOp1ewH5CQF1OvZtZMQZt9njcMTitsR64ESgBMg3M64SAnH8A0wjryJ8zs1WSqrLyLiJEd04gOhXCOu4jZrZa0oLsgiQNB/5qZt+U9Gg8fAewk9AgVmdpuAb4RWx8VhAant1mdn8z/fuBCuBmQqN3DXA1UAnsiefKCQ3VSWCspLfHvPXAf4AHs8otAw5YXOdtZqdDG8vHJS0DzpjZHKBE0veBq4BxMW9VTANh/fo3gPsklQIrMrEQZnY49tjrCFs+1BKC0G7MlCtpnJlNkjSG0JADrDOzHZI2JdTZWaJT+1O0h+icd5vZ/ZLGEhr/l4DvAc/GbB+Mr3VmdkjSdkLU98lYDoSYlBckPSnpBeC0mc2Lmq9tod73N9fmpI87Bqc1TsbXvxP2dIHQY15ICBD8VnzNBMRkXk+ZWZOkM4QeYoamrPevA71i77VvvM7prHMQpjvvNbM3munKLrM1thNGOOsJjuFNwA4zWwwg6W3ACDO7JY4s+sV8Jwi932z9/+RcFGtm5ALNRgzACTObJekTwHXASpqNGICjwByFfYs2RJ0AmNkaYI2kKcCnWrC1eX1DGD1Bcp2dJdb5pdGexGTAITNbkJVvUlY5J5udW0Cos0xekXu/c+rdKVzcMThtIROdeXP8/HPgq4QRA4Qeab2kywjTKqfyXOv3wFcknSBs4DYH2E0ItnxJ0ucl3UnYsgHCFEW9pMPAi7HBzJS5LI4yDpjZ0dhzb4m1wI8516A/G/MuBN5KmIrpJ2kuYdokkdjTXippFWE7lVOEqNbMiAGy9r0xs03xd4YnyR0xPATcQIjcvYDgGACQdBHB8R4nTNfNI0yvLZf0b8JIbJukxYQpni8TtrfIkFRnAB+T9DBh+mdJtCfJ1mOSfi1pCaGRf7xZktWSlhM6D5tbyP87SX1jPb8CLCO33ueYWb7/FSclPPLZ6RQk3QW8EzhqZt9OW4/jOB3HHYPjOI6Tgy85cxzHcXJwx+A4juPk4I7BcRzHycEdg+M4jpODOwbHcRwnB3cMjuM4Tg7/BcKA3ZoZVu22AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Axis breaking method taken from https://matplotlib.org/examples/pylab_examples/broken_axis.html\n",
    "\n",
    "f, (ax, ax2) = plt.subplots(2, 1, sharex=True, gridspec_kw={'height_ratios': [1, 3]})\n",
    "\n",
    "# plot the same data on both axes\n",
    "ax.scatter(plot_negative[\"differences\"], plot_negative[\"logp\"], \n",
    "           c=plot_negative[\"color\"], alpha=0.5, s=3)\n",
    "ax2.scatter(plot_negative[\"differences\"], plot_negative[\"logp\"],\n",
    "            c=plot_negative[\"color\"], alpha=0.5, s=3)\n",
    "\n",
    "# zoom-in / limit the view to different portions of the data\n",
    "ax.set_ylim(33, 36)  # outliers only\n",
    "ax2.set_ylim(0, 12)  # most of the data\n",
    "\n",
    "# hide the spines between ax and ax2\n",
    "ax.spines['bottom'].set_visible(False)\n",
    "ax2.spines['top'].set_visible(False)\n",
    "ax.xaxis.tick_top()\n",
    "ax.tick_params(labeltop=False)  # don't put tick labels at the top\n",
    "ax2.xaxis.tick_bottom()\n",
    "\n",
    "d = .015  # how big to make the diagonal lines in axes coordinates\n",
    "# arguments to pass to plot, just so we don't keep repeating them\n",
    "kwargs = dict(transform=ax.transAxes, color='k', clip_on=False)\n",
    "ax.plot((-d, +d), (-d, +d*5), **kwargs)        # top-left diagonal\n",
    "ax.plot((1 - d, 1 + d), (-d, +d*5), **kwargs)  # top-right diagonal\n",
    "\n",
    "kwargs.update(transform=ax2.transAxes)  # switch to the bottom axes\n",
    "ax2.plot((-d, +d), (1 - d, 1 + d), **kwargs)  # bottom-left diagonal\n",
    "ax2.plot((1 - d, 1 + d), (1 - d, 1 + d), **kwargs)  # bottom-right diagonal\n",
    "\n",
    "f.subplots_adjust(hspace=0.1)\n",
    "\n",
    "texts=[]\n",
    "\n",
    "# Naming select points\n",
    "labeled_points = [\"FOSL1\"]\n",
    "\n",
    "ax.annotate(\"KRAS\", (plot_negative[plot_negative[\"genes\"] == \"KRAS\"][\"differences\"]-0.05,\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"KRAS\"][\"logp\"]-1.7),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"MYH9\", (plot_negative[plot_negative[\"genes\"] == \"MYH9\"][\"differences\"],\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"MYH9\"][\"logp\"]+0.1),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"TEAD1\", (plot_negative[plot_negative[\"genes\"] == \"TEAD1\"][\"differences\"]-.02,\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"TEAD1\"][\"logp\"]-0.3),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"RAF1\", (plot_negative[plot_negative[\"genes\"] == \"RAF1\"][\"differences\"]-0.02,\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"RAF1\"][\"logp\"]-0.1),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"DOCK5\", (plot_negative[plot_negative[\"genes\"] == \"DOCK5\"][\"differences\"]-0.06,\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"DOCK5\"][\"logp\"]-0.2),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"BCAR1\", (plot_negative[plot_negative[\"genes\"] == \"BCAR1\"][\"differences\"]-0.07,\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"BCAR1\"][\"logp\"]-0.85),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"CTNNB1\", (plot_negative[plot_negative[\"genes\"] == \"CTNNB1\"][\"differences\"],\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"CTNNB1\"][\"logp\"]-0.9),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"CFLAR\", (plot_negative[plot_negative[\"genes\"] == \"CFLAR\"][\"differences\"]+0.01,\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"CFLAR\"][\"logp\"]-0.4),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"TCF7L2\", (plot_negative[plot_negative[\"genes\"] == \"TCF7L2\"][\"differences\"]+0.01,\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"TCF7L2\"][\"logp\"]-0.4),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"RAB10\", (plot_negative[plot_negative[\"genes\"] == \"RAB10\"][\"differences\"],\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"RAB10\"][\"logp\"]-0.6),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"CRK\", (plot_negative[plot_negative[\"genes\"] == \"CRK\"][\"differences\"]-0.01,\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"CRK\"][\"logp\"]+0.3),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"TLN1\", (plot_negative[plot_negative[\"genes\"] == \"TLN1\"][\"differences\"],\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"TLN1\"][\"logp\"]),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none',  edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "ax2.annotate(\"FOSL1\", (plot_negative[plot_negative[\"genes\"] == \"FOSL1\"][\"differences\"]+0.02,\n",
    "                  plot_negative[plot_negative[\"genes\"] == \"FOSL1\"][\"logp\"]-0.5),\n",
    "                textcoords='offset points',\n",
    "                bbox=dict(boxstyle='round,pad=0.2', facecolor='none', edgecolor=\"black\"),\n",
    "                fontsize=8)\n",
    "\n",
    "plt.xlabel(\"Magnitude of Mean CERES Score Difference\", fontsize=7)\n",
    "plt.xticks(fontsize=7)\n",
    "plt.yticks(fontsize=7)\n",
    "ax.tick_params(labelsize=7)\n",
    "f.text(0.04, 0.5, \"-log10(p value)\", va='center', rotation='vertical', fontsize=7)\n",
    "plt.savefig(\"../output/FigureB_V2_MeanCERESScoreDifference_NegOnly_WithAxesBreak.pdf\")\n",
    "\n",
    "ax.annotate(\"KRAS mutant vs. WT cell lines\", (0.15, 35), fontsize=8)\n",
    "ax.annotate(\"Significant (p < 0.05)\", (0.15, 34.3), c=\"r\", fontsize=8)\n",
    "plt.savefig(\"../output/FigureB_V2_MeanCERESScoreDifference_NegOnly_WithAxesBreak_withPlotText.pdf\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RNAi Values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-142-d406893b34ca>:1: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  depmap.rnai.rnai_data.sort_values(\"KRAS\")[\"KRAS\"].to_csv(\"../output/KRAS_RNAi_Dependency_Values.txt\",\n"
     ]
    }
   ],
   "source": [
    "depmap.rnai.rnai_data.sort_values(\"KRAS\")[\"KRAS\"].to_csv(\"../output/KRAS_RNAi_Dependency_Values.txt\",\n",
    "                                                         sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dependency Stratification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras_dependency_scores = pd.DataFrame(depmap.rnai.rnai_data.sort_values(\"KRAS\")[\"KRAS\"]).reset_index()\n",
    "kras_dependency_scores.rename(columns = {\"index\": \"CCLE_Name\"}, inplace=True)\n",
    "\n",
    "# Adding DepMap_ID to CCLE names\n",
    "sample_info_dict = sample_info.set_index(\"CCLE_Name\").to_dict()\n",
    "kras_dependency_scores[\"DepMap_ID\"] = kras_dependency_scores[\"CCLE_Name\"].map(sample_info_dict[\"DepMap_ID\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras_depend = kras_dependency_scores[kras_dependency_scores[\"KRAS\"] <= -0.5].dropna()\n",
    "kras_independ = kras_dependency_scores[kras_dependency_scores[\"KRAS\"] > -0.5].dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras_depend_achilles = depmap.achilles.achilles_data.loc[kras_depend[\"DepMap_ID\"], :].dropna()\n",
    "kras_independ_achilles = depmap.achilles.achilles_data.loc[kras_independ[\"DepMap_ID\"], :].dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_ceres_depend = kras_depend_achilles.mean(axis=0)\n",
    "mean_ceres_independ = kras_independ_achilles.mean(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "dep_ind_test_results = scipy.stats.ttest_ind(kras_depend_achilles, kras_independ_achilles, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_depend_difference = mean_ceres_depend - mean_ceres_independ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = {\"genes\": mean_depend_difference.index, \"differences\": mean_depend_difference.values, \"logp\": -np.log10(dep_ind_test_results[1])}\n",
    "\n",
    "plot_df = pd.DataFrame(data)\n",
    "plot_df.to_csv(\"../output/KRAS_Dependency_Group_Differences_Volcano.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating Correlations of Dependencies with BCAR1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Calculate pearson correlation between each gene CERES profiles.\n",
    "2. Calculate Z-Scores of pearson correlations.\n",
    "3. Z-Score > 3 can be considered correlated.\n",
    "\n",
    "Source: https://depmap.org/rnai/data/index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "achilles_data = depmap.achilles.achilles_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcar1_corr = achilles_data.corrwith(achilles_data[\"BCAR1\"], axis=0, method=\"pearson\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcar1_corr_zscore = (bcar1_corr - bcar1_corr.mean())/(bcar1_corr.std())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "corr_genes = bcar1_corr_zscore[bcar1_corr_zscore > 3].index.to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Printing out the highly correlated genes for input into STRING."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACTR2\n",
      "ACTR3\n",
      "AFDN\n",
      "ALG14\n",
      "ALG6\n",
      "ANKRD11\n",
      "ARHGEF12\n",
      "ARHGEF7\n",
      "ARPC2\n",
      "ARPC3\n",
      "ARPC4\n",
      "ATRN\n",
      "BCAR1\n",
      "BCAR3\n",
      "BRK1\n",
      "BTBD3\n",
      "CCDC113\n",
      "CDC42\n",
      "CDH2\n",
      "CRK\n",
      "CRKL\n",
      "CTRB1\n",
      "DLG5\n",
      "DOCK5\n",
      "DOCK7\n",
      "DYNC1LI2\n",
      "EFNA5\n",
      "ELMO2\n",
      "ELOVL1\n",
      "FERMT2\n",
      "FGFR1OP2\n",
      "FNDC3B\n",
      "FOSL1\n",
      "GGNBP2\n",
      "GLG1\n",
      "GNB1\n",
      "HSP90B1\n",
      "ILK\n",
      "ITGAV\n",
      "ITGB5\n",
      "KLHL20\n",
      "KRAS\n",
      "LIMS1\n",
      "LPAR1\n",
      "MARK2\n",
      "MARK3\n",
      "MESD\n",
      "MGRN1\n",
      "MOB4\n",
      "MYO10\n",
      "NCKAP1\n",
      "OST4\n",
      "PAK2\n",
      "PAK4\n",
      "PGAP2\n",
      "PHACTR4\n",
      "PIGT\n",
      "PKN2\n",
      "PPP2R5C\n",
      "PTK2\n",
      "PTPN1\n",
      "PXN\n",
      "RAC1\n",
      "RAP1GDS1\n",
      "RAPGEF1\n",
      "RAPGEF6\n",
      "RAPH1\n",
      "RBBP7\n",
      "RHOG\n",
      "RIC8A\n",
      "RSAD1\n",
      "RSU1\n",
      "RTCB\n",
      "SEPT9\n",
      "SETD5\n",
      "SLC12A3\n",
      "SLCO1C1\n",
      "SRC\n",
      "STT3A\n",
      "TEAD1\n",
      "TEAD4\n",
      "TIPARP\n",
      "TLN1\n",
      "TM2D2\n",
      "TM2D3\n",
      "TNS3\n",
      "TP53BP2\n",
      "TRIO\n",
      "TSPAN4\n",
      "TULP4\n",
      "UBIAD1\n",
      "VASP\n",
      "VCL\n",
      "WWTR1\n",
      "YES1\n",
      "ZFP36L1\n"
     ]
    }
   ],
   "source": [
    "print('\\n'.join(corr_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-93-97aa11a5e8e9>:1: FutureWarning: The signature of `Series.to_csv` was aligned to that of `DataFrame.to_csv`, and argument 'header' will change its default value from False to True: please pass an explicit value to suppress this warning.\n",
      "  bcar1_corr_zscore.sort_values(ascending=False).to_csv(\"../output/BCAR1_Correlation_ZScores.txt\", sep=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "bcar1_corr_zscore.sort_values(ascending=False).to_csv(\"../output/BCAR1_Correlation_ZScores.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BCAR1\n",
      "CRKL\n",
      "PTK2\n",
      "ILK\n",
      "RAC1\n",
      "TLN1\n",
      "ITGAV\n",
      "CRK\n",
      "LIMS1\n",
      "NCKAP1\n",
      "RSU1\n",
      "DOCK5\n",
      "ELMO2\n",
      "PKN2\n",
      "RAPH1\n",
      "MYO10\n",
      "VCL\n",
      "WWTR1\n",
      "TRIO\n",
      "ITGB5\n",
      "PXN\n",
      "FERMT2\n",
      "RAPGEF1\n",
      "MOB4\n",
      "SRC\n"
     ]
    }
   ],
   "source": [
    "print('\\n'.join(bcar1_corr_zscore.sort_values(ascending=False).head(25).index.to_list()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## KRAS Dependent Pancreatic Lines vs. Non-Pancreatic KRAS Independent"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This data analysis follows the same method used for the KRAS dependent vs independent comparison, except that the the two groups being used for comparison are:\n",
    "    1. KRAS dependent pancreatic lines\n",
    "    2. Non-Pancreatic KRAS independent lines\n",
    "    \n",
    "The same dependency stratification can be used for identifying lines, since full CCLE names have the tissue name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras_depend_panc = kras_depend[kras_depend[\"CCLE_Name\"].str.contains(\"PANCREAS\")]\n",
    "kras_independ_non_panc = kras_independ[~kras_independ[\"CCLE_Name\"].str.contains(\"PANCREAS\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras_depend_panc = depmap.achilles.achilles_data.loc[kras_depend_panc[\"DepMap_ID\"], :].dropna()\n",
    "kras_independ_non_panc = depmap.achilles.achilles_data.loc[kras_independ_non_panc[\"DepMap_ID\"], :].dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_ceres_panc = kras_depend_panc.mean(axis=0)\n",
    "mean_ceres_non_panc = kras_independ_non_panc.mean(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "dep_ind_test_results_panc = scipy.stats.ttest_ind(kras_depend_panc, kras_independ_non_panc, equal_var=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_depend_difference_panc = mean_ceres_panc - mean_ceres_non_panc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = {\"genes\": mean_depend_difference_panc.index, \"differences\": mean_depend_difference_panc.values, \"logp\": -np.log10(dep_ind_test_results_panc[1])}\n",
    "\n",
    "plot_df = pd.DataFrame(data)\n",
    "plot_df.to_csv(\"../output/KRAS_PancDependency_NonPancKRASIndependGroup_Differences.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STRING Network Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "import networkx as nx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "string_network = pd.read_csv(\"../output/Top25_BCAR1_Correlations_string_interactions-2.tsv\", sep=\"\\t\")\n",
    "\n",
    "edges = []\n",
    "string_nodes = set(string_network[\"node1\"].tolist()).union(\n",
    "    set(string_network[\"node2\"].tolist()))\n",
    "\n",
    "for e in range(string_network.shape[0]):\n",
    "    edges.append((string_network.at[e, \"node1\"],\n",
    "                  string_network.at[e, \"node2\"],\n",
    "                  {\"weight\": float(string_network.at[e, \"combined_score\"])}))\n",
    "    \n",
    "string_graph = nx.Graph()\n",
    "string_graph.add_nodes_from(string_nodes)\n",
    "string_graph.add_edges_from(edges)\n",
    "nx.write_gml(string_graph, \"../output/Top25_BCAR1Corr_StringGML.gml\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
